algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs

```diff
@@ -21,249 +20,286 @@ use crate::{
-                    .map(|a| v_H_i_alpha_v_H_i_beta * a)
-                    .collect();
-                DensePolynomial::from_coefficients_vec(coeffs)
-            };
-            end_timer!(a_poly_time);
-            a_poly
-        });
-
-        let (row_on_K, col_on_K, row_col_on_K) =
-            (&arithmetization.evals_on_K.row, &arithmetization.evals_on_K.col, &arithmetization.evals_on_K.row_col);
-
-        job_pool.add_job(|| {
-            let b_poly_time = start_timer!(|| format!("Computing b poly for {label}"));
-            let alpha_beta = alpha * beta;
-            let b_poly = {
-                let evals: Vec<F> = cfg_iter!(row_on_K.evaluations)
-                    .zip_eq(&col_on_K.evaluations)
-                    .zip_eq(&row_col_on_K.evaluations)
-                    .map(|((r, c), r_c)| alpha_beta - alpha * r - beta * c + r_c)
-                    .collect();
-                EvaluationsOnDomain::from_vec_and_domain(evals, non_zero_domain)
-                    .interpolate_with_pc(ifft_precomputation)
-            };
-            end_timer!(b_poly_time);
-            b_poly
-        });
-        let [a_poly, b_poly]: [_; 2] = job_pool.execute_all().try_into().unwrap();
-
-        let f_evals_time = start_timer!(|| format!("Computing f evals on K for {label}"));
-        let mut inverses: Vec<_> = cfg_iter!(row_on_K.evaluations)
-            .zip_eq(&col_on_K.evaluations)
-            .map(|(r, c)| (beta - r) * (alpha - c))
+        assignment: DensePolynomial<F>,
+        arith: &MatrixArithmetization<F>,
+        c_i_to_k: BTreeMap<usize, Vec<usize>>,
+        alpha: F,
+        combiner: F,
+    ) -> Result<LinevalInstance<F>, anyhow::Error> {
+        let sumcheck_time = start_timer!(|| format!("Compute LHS of sumcheck for {_label}"));
+
+        let v_R: SparsePolynomial<F> = constraint_domain.vanishing_polynomial();
+
+        let m_at_alpha_evals_time = start_timer!(|| format!("Compute m_at_alpha_evals parallel for {_label}"));
+        let mut m_at_alpha_evals: BTreeMap<usize, F> = cfg_iter!(c_i_to_k)
+            .map(|(&c_i, ks)| {
+                let mut m_at_c_i = F::zero();
+                for &k in ks {
+                    let col_at_k = arith.evals_on_K.col[k];
+                    let row_at_k = arith.evals_on_K.row[k];
+                    let row_col_val_at_k = arith.evals_on_K.row_col_val[k];
+                    let l_row_at_k = constraint_domain.size_inv * v_R.evaluate(alpha) / (alpha - row_at_k);
+                    // instead of calculating L_col(k), we just add val(k)*L_row(k) directly to the evaluation where we know L_col(k) will be 1
+                    // recall that row_col_val(k) = M_{i,j}*row(k)*col(k). Given that we 'ignore' L_col(k), we divide out the col(k) factor
+                    m_at_c_i += row_col_val_at_k * l_row_at_k / col_at_k;
+                }
+                (c_i, m_at_c_i)
+            })
```

Comment:
We don't need to compute the evaluations of M(\alpha, c) for all c \in C manually like this. Instead, let's use the following strategy: compute the transpose of the matrix (in the offline phase? but it's fine to do it online too), and then, we can iterate as follows:
```rust
for col in matrix_transpose {
    let mut sum_for_col = F::zero();
    for (val, row_index) in col {
        sum_for_col += val * L_row(alpha)
    }
}
```
We can precalculate L_row(alpha) outside the loop too (`constraint_domain.evaluate_all_lagrange_polynomials(alpha)`)
This should be simpler to understand, and cleaner to audit.

It would also be nice to have some comments indicating what this code block is doing.

Reply:
That is a lot cleaner indeed, see: a161a6d73 and 81d98173f
No need to precompute in the offline phase as for a "normal" transfer credit proof, `calc_transposes` takes less than a percentage of the 3rd round's time.